### DISCUSSION TITLE
systemd Vector Service Configuration with Environment Variables

### DISCUSSION THREAD

**Discussion by tbenade** [vectordotdev/vector#20141] [Q&A] [answered]
Hi,

I was hoping to use some templated configuration only swapping out the value of environment variables  in `/etc/default/vector` on Linux EC2 instances running vector as a systemd service. We already used `/etc/default/vector` to configure some `VECTOR_*` values.

Deployed a quick POC and ran great. Then noticed in the `validate` output that we have a deprecation warning.

```
Loaded with warnings ["/etc/vector"]
------------------------------------
~ Unknown environment variable in config. This is DEPRECATED and will become an error in future versions. name = "SPLUNK_INDEX"
```
Two questions then. Maybe this guidance may be misleading for anyone comes across it. https://github.com/vectordotdev/vector/blob/master/distribution/systemd/vector.default#L2

Next question is what is the appropriate mechanism for despatching environment variables for a vector systemd service if not `/etc/default/vector` please. Apologies if this is purely a linux question, systemd is not in my regular skillset.

Thanks

**Comment by jszwedko** [comment]
`/etc/environment/vector` will automatically be loaded by the default SystemD unit: https://github.com/vectordotdev/vector/blob/58a4a2ef52e606c0f9b9fa975cf114b661300584/distribution/systemd/vector.service#L16. Do you have a line like `SPLUNK_INDEX=some_index` in it?

**Reply by tbenade** [reply]
Correct the file is loaded and variables in the file are correctly loaded. My question is why the deprecation notice for "unknown" (custom variables) when running `validate`. Is the `EnvironmentFile` not the best mechanism for defining systemd environment variables?

**Reply by jszwedko** [reply, marked as answer]
Correct, undefined variables in Vector configuration will become a hard error in the future. You can provide a default to preserve the current behavior with bash-like syntax. For example: `${SPLUNK_INDEX-}`.

**Reply by tbenade** [reply]
Thanks for that. I'm building new so not looking to introduce a regression. Where else does one define environment variables for vector aggregator running as a service if we can't use an `EnvironmentFile`. Thanks

**Reply by tbenade** [reply]
Might try answer my own question in an attempt to further my own learning.

I think I need to create an overrides service unit file which will point to an additional `EnvironmentFile` where I will be able to set my environment variables without upsetting `validate`?

**Reply by tbenade** [reply]
Yikes, I now realise I totally misunderstood. The error is not being generated by the loading of the systemd unit file but the parsing of my vector configuration 😵‍💫 . Sorry for being so confusing. In the end my issue was with the fact in my vector configuration in this case the Splunk HEC configuration I was referencing an "unknown" environment variable.

Thanks for the guiding me back to sanity.

**Reply by tbenade** [reply]
Documentation clearly explains Jesse's suggestion, https://vector.dev/docs/reference/configuration/#default-values

**Reply by tbenade** [reply]
@jszwedko you don't happen to have a suggestion how I can change/reload environment variables for vector as a service? I thought I would be able to have a deployment process replace the `EnviromentFile` then restart the service unit and reload my new settings but that is not the case for systemd at all. 🤷

**Reply by jszwedko** [reply]
It looks like systemd doesn't reload the `EnvironmentFile` when using `reload` it so you'll need to use `restart` instead.
